### ISSUE TITLE
why no variable 'pdpath'?

### ISSUE THREAD

**Issue by katjav** [pure-data/pd-lib-builder#26]
The Makefile Template of pd-extended defines variable 'PD_PATH', from where 'PD_INCLUDE' (for pd API files) is derived per platform, plus paths for binaries (for Windows only).

Makefile.pdlibbuilder only has 'pdincludepath' for the directory where pd API files should sit, and 'pdbinpath' for the directory where pd.dll should be found on Windows. It doesn't define a root path for pd components, which is particularly inconvenient when you have to type two Windows paths on command line. Why not introduce 'pdpath'?

The thing is, a root path could only work if a standardized directory tree is defined for all pd flavors and build situations. And that is not at all the case. A portable install as used in a centralized build differs from a regular install. Standard pd install locations vary not only per platform (which we can detect), but there's also layout variations per pd flavor. All together it seems impossible to define just one variable and derive appropriate paths from it for all cases. Yet I'd like to make things a bit more convenient. I don't know how yet, this is just brainstorming. Let's start with an overview of standard paths per pd flavor and per platform:

vanilla pd regular installation:

```
Linux: /usr/include/pd
OSX: /Applications/Pd*.app/Contents/Resources/src
Windows: %PROGRAMFILES%/pd/src
Windows: %PROGRAMFILES%/pd/bin
```

pd-extended regular installation:

```
Linux: /usr/include/pdextended
OSX: /Applications/Pd-extended*.app/Contents/Resources/include/pdextended
Windows: %PROGRAMFILES%/pd/include/pdextended
Windows: %PROGRAMFILES%/pd/bin
```

centralized build:

```
Linux: $(PD_PATH)/include/pd
OSX: $(PD_PATH)/include/pd
Windows: $(PD_PATH)/include/pd, $(PD_PATH)/bin
```

Linux:

```
vanilla pd regular installation: /usr/include/pd
pd-extended regular installation: /usr/include/pdextended
centralized build: $(PD_PATH)/include/pd
```

OSX:

```
vanilla pd regular installation: /Applications/Pd*.app/Contents/Resources/src
pd-extended regular installation: /Applications/Pd-extended*.app/Contents/Resources/include/pdextended
centralized build: $(PD_PATH)/include/pd

```
Windows:

```
vanilla pd regular installation: %PROGRAMFILES%/pd/src
vanilla pd regular installation: %PROGRAMFILES%/pd/bin
pd-extended regular installation: %PROGRAMFILES%/pd/include/pdextended
pd-extended regular installation: %PROGRAMFILES%/pd/bin
centralized build: $(PD_PATH)/include/pd
centralized build: $(PD_PATH)/bin
```

Pd-l2ork is another flavor to reckon with. It is the active fork of pd-extended and has inherited the build system. Many makefiles in the pd-l2ork repositories still speak of 'Pd-extended' but it installs under its own application name nowadays. There's no pd-lib-builder in pd-l2ork yet but some day they may want to update actively maintained libs that use it.

Well I don't know what to make of this now. To be continued.

**Comment by katjav** [comment]
The path to dll is only needed on Windows. That makes it easier, maybe it could be calculated from the include path so you only need to define that one.

**Comment by katjav** [comment]
To further confuse things, my overview in the first post doesn't display correct paths for the centralized build case, and moreover they are not everywhere consistent between pd-extended and pd-l2ork. For example, the externals root makefile in pd-extended calls make for lib hcs like so:

```
hcs:
    make -C $(externals_src)/hcs CFLAGS="$(CFLAGS)" \
		PD_PATH=$(pd_src) PD_INCLUDE=$(pd_src)/src
```

Variable definitions in these make command arguments override corresponding ones in the hcs lib makefile. The same section in purr-data / pd-l2ork doesn't pass a definition for PD_INCLUDE:

```
hcs:
    make -C $(externals_src)/hcs PD_PATH=$(pd_src) CFLAGS="$(CFLAGS)"
```

Meaning it leaves the definition of PD_INCLUDE to the hcs lib makefile, where it is different:

```
PD_INCLUDE = $(PD_PATH)/include/pd
```

A path like this is not the source tree, maybe it is created in the build. In contrast, for the flatgui lib, pd-l2ork does pass the include path:

```
flatgui:
    make -C $(externals_src)/flatgui CFLAGS="$(CFLAGS)" \
        PD_PATH=$(pd_src) PD_INCLUDE=$(pd_src)/src
```

Anyhow, from these examples we see that a single argument  PD_PATH=\<pd-rootdir\> doesn't do the job because PD_INCLUDE is derived from pd_src in the externals root makefile, not from PD_PATH.

For a Makefile.pdlibbuilder-based lib the corresponding arguments passed from an externals root makefile would be:

```    
 pdbinpath=$(pd_src)/bin pdincludepath=$(pd_src)/src
```

Now it looks like we could support variable 'PD_PATH' and define in Makefile.pdlibbuilder:

pdincludepath = $(PD_PATH)/src
pdbinpath = $(PD_PATH)/bin

But... while this would only work when building against a 'portable' pd install (including centralized builds), not with regular installed packages in general. Is that bad? It seems that it should work in all cases where a root dir literally called 'pd' exists.

**Comment by katjav** [comment]
Implemented the above with commit 41e9743. Although a small addition I bumped version to 0.3.0 to reflect 'new feature', because the change also completes compatibility with the existing central makefile in extended / l2ork, as far as template-based libs are concerned.

It is important to know that PD_PATH points to the root dir of a 'portable' pd installation: a pd directory containing all components, notably bin/ and src/. For regular installs, subdir paths differ per pd flavor and platform so there you would use pdincludepath to point directly to the includes path (and pdbinpath for the dir with pd.dll on Windows)

PD_PATH can also be set in the environment. In a central makefile you could set:

export PD_PATH := <path-to-pd-rootdir>

and all submake processes will get the definition so you don't need to specify per submake as command argument. You could also define PD_PATH in a config makefile for personal use, be sure not to distribute such a thing with your lib source tree. If the name is Makefile.config and it sits in the parent of the lib tree, in your lib makefile you would state:

-include ../Makefile.config

The dash is important, to indicate it is optional and should not generate an error when not found.

**Comment by katjav** [comment]
While variable 'PD_PATH' is useful for compatibility with central makefile (inherited) from pd-extended, there's a more consistent set of path variables introduced with version 0.4.0, see issue #27.
